src/snowflake/snowpark/_internal/compiler/utils.py

```diff
@@ -58,26 +62,71 @@ def create_query_generator(plan: SnowflakePlan) -> QueryGenerator:
     return QueryGenerator(plan.session, snowflake_create_table_plan_info)
 
 
-def replace_child(
+def resolve_and_update_snowflake_plan(
+    node: SnowflakePlan, query_generator: QueryGenerator
+) -> None:
+    """
+    Re-resolve the current snowflake plan if it has a source plan attached, and update the fields with
+    newly resolved value.
+    """
+
+    if node.source_plan is None:
+        return
+
+    new_snowflake_plan = query_generator.resolve(node.source_plan)
+
+    # copy over the newly resolved fields to make it an in-place update
+    node.queries = new_snowflake_plan.queries
+    node.post_actions = new_snowflake_plan.post_actions
+    node.expr_to_alias = new_snowflake_plan.expr_to_alias
+    node.is_ddl_on_temp_object = new_snowflake_plan.is_ddl_on_temp_object
+    node._output_dict = new_snowflake_plan._output_dict
+    node.df_aliased_col_name_to_real_col_name = (
+        new_snowflake_plan.df_aliased_col_name_to_real_col_name
+    )
+    node.placeholder_query = new_snowflake_plan.placeholder_query
+    node.referred_ctes = new_snowflake_plan.referred_ctes
+    node._cumulative_node_complexity = new_snowflake_plan._cumulative_node_complexity
+
+
+def resolve_node(node: LogicalPlan, query_generator: QueryGenerator) -> SnowflakePlan:
+    resolved_node = query_generator.resolve(node)
+    resolved_node._is_valid_for_replacement = True
```

Comment:
should we add this inside query_generator since we know this snowflakeplan will be generated during optimization stage?

Reply:
that is a good idea, i moved this to query generator resolve_with_children